src/features/tokens/adapters/SealevelTokenAdapter.ts

```diff
@@ -0,0 +1,416 @@
+    const tx = new Transaction({
+      feePayer: fromWalletPubKey,
+      blockhash: recentBlockhash,
+      recentBlockhash,
+    }).add(transferRemoteInstruction);
+    tx.partialSign(randomWallet);
+    return tx;
+  }
+
+  // Should match https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/trevor/sealevel-validator-rebase/rust/sealevel/libraries/hyperlane-sealevel-token/src/processor.rs#LL49C1-L53C30
+  deriveHypTokenAccount(): PublicKey {
+    const [pda] = PublicKey.findProgramAddressSync(
+      [
+        Buffer.from('hyperlane_message_recipient'),
+        Buffer.from('-'),
+        Buffer.from('handle'),
+        Buffer.from('-'),
+        Buffer.from('account_metas'),
+      ],
+      this.warpProgramPubKey,
+    );
+    return pda;
+  }
+
+  // https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/trevor/sealevel-validator-rebase/rust/sealevel/programs/mailbox/src/pda_seeds.rs#L19
+  deriveMailboxOutboxAccount(mailbox: PublicKey): PublicKey {
+    const [pda] = PublicKey.findProgramAddressSync(
+      [Buffer.from('hyperlane'), Buffer.from('-'), Buffer.from('outbox')],
+      mailbox,
+    );
+    return pda;
+  }
+
+  // https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/trevor/sealevel-validator-rebase/rust/sealevel/programs/mailbox/src/pda_seeds.rs#L57
+  deriveMessageDispatchAuthorityAccount(): PublicKey {
+    const [pda] = PublicKey.findProgramAddressSync(
+      [Buffer.from('hyperlane_dispatcher'), Buffer.from('-'), Buffer.from('dispatch_authority')],
+      this.warpProgramPubKey,
+    );
+    return pda;
+  }
+
+  // https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/trevor/sealevel-validator-rebase/rust/sealevel/programs/mailbox/src/pda_seeds.rs#L33-L37
+  deriveMsgStorageAccount(mailbox: PublicKey, randomWalletPubKey: PublicKey): PublicKey {
+    const [pda] = PublicKey.findProgramAddressSync(
+      [
+        Buffer.from('hyperlane'),
+        Buffer.from('-'),
+        Buffer.from('dispatched_message'),
+        Buffer.from('-'),
+        randomWalletPubKey.toBuffer(),
+      ],
+      mailbox,
+    );
+    return pda;
+  }
+
+  // https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/trevor/sealevel-validator-rebase/rust/sealevel/programs/hyperlane-sealevel-token-native/src/plugin.rs#L26
+  deriveNativeTokenCollateralAccount(): PublicKey {
+    const [pda] = PublicKey.findProgramAddressSync(
+      [Buffer.from('hyperlane_token'), Buffer.from('-'), Buffer.from('native_collateral')],
+      this.warpProgramPubKey,
+    );
+    return pda;
+  }
+}
+
+// Interacts with Hyp Collateral token programs
+export class SealevelHypCollateralAdapter extends SealevelHypNativeAdapter {
+  override getTransferInstructionKeyList(
+    sender: PublicKey,
+    mailbox: PublicKey,
+    randomWallet: PublicKey,
+  ): Array<AccountMeta> {
+    const commonKeys = super
+      .getTransferInstructionKeyList(sender, mailbox, randomWallet)
+      .slice(0, 9);
```

Comment:
I def don't feel strongly here but noting if the setup was an abstract `SealevelHypAdapter` as the superclass that just contained the common keys in `getTransferInstructionKeyList`, then `SealevelHypNativeAdapter` could also be a subclass of that and each of the subclasses could just `const commonKeys = super.getTransferInstructionKeyList(sender, mailbox, randomWallet)` without being responsible for slicing to get the common keys. Just noting it could be slightly less fragile if the common keys were to change in length or something, but happy to keep as is

Reply:
Will do thanks